openstack-auto-scaling.sh

```diff
@@ -0,0 +1,349 @@
+    if test -r "$LOCKFILE" -a -f "$LOCKFILE" ; then 
+      log "2" "Lock file $LOCKFILE exist, script is not going to take any action"
+      exit 0
+    fi
+  else
+    log "2" "Lock file override enabled"
+  fi
+}
+
+get_agg_group_status() {
+#usage: get_agg_group_status AGG_GROUP
+  local agg_group="$1"
+  openstack_overcloud aggregate show "$agg_group" &>/dev/null
+  local RETVAL
+  RETVAL="$?"
+  if [ "$RETVAL" -ne 0 ]; then
+    log "1" "Unable to find aggregation group $agg_group" && return 1
+  else
+    local agg_group_hypervisor_list 
+    agg_group_hypervisor_list="$(openstack_overcloud aggregate show "$agg_group" -f json -c hosts |jq --raw-output '.hosts[]')"
+    local agg_group_hypervisor_filter
+    agg_group_hypervisor_filter="$(echo "$agg_group_hypervisor_list"|tr '\n' '|'|sed 's/|$//g')"
+
+    local hyp_list
+    hyp_list="$(openstack_overcloud hypervisor list --long -f value -c "Hypervisor Hostname" -c "vCPUs Used" -c "State"| \
+                  grep -E "$agg_group_hypervisor_filter"|sort)"
+
+    computes_number="$(echo "$hyp_list"|wc -l)"
+    computes_number_up_with_vm="$(echo "$hyp_list"|grep " up "|grep -vc " 0$")"
+    computes_number_up_before_logic="$(echo "$hyp_list"|grep -c " up ")"
+    computes_number_up_after_logic="$((computes_number_up_with_vm + computes_number * UCUPN / 100))"
+    [ "$computes_number_up_after_logic" -gt "$computes_number" ] && \
+      computes_number_up_after_logic="$computes_number"
+
+    computes_number_difference="$((computes_number_up_before_logic - computes_number_up_after_logic))"
+
+    computes_hostnames_up="$(echo "$hyp_list"|grep " up "|awk '{print $1}')"
+    computes_hostnames_up_available_for_shutdown="$(echo "$hyp_list"|grep " up "|grep " 0$"|awk '{print $1}')"
+    computes_hostnames_down_with_no_vm="$(echo "$hyp_list"|grep " down "|grep " 0$"|awk '{print $1}')"
+
+    local computes_number_down_with_vm
+    computes_number_down_with_vm="$(echo "$hyp_list"|grep " down "|grep -vc " 0$")"
+    local computes_hostnames_down_with_vm
+    computes_hostnames_down_with_vm="$(echo "$hyp_list"|grep " down "|grep -v " 0$"|awk '{print $1}')"
+    [ "$computes_number_down_with_vm" -ne 0 ] && \
+      log "2" "IMPORTANT: hypervisors in aggregation group $agg_group down with VMs: $(echo $computes_hostnames_down_with_vm)"
+  fi
+  return 0
+}
+
+print_agg_group_status() {
+#function expects global variables
+  log "0" "Percentage of empty computes configured: $UCUPN%"
+  log "0" "Number of hypervisors in aggregation group $agg_group: $computes_number"
+  log "0" "Number of hypervisors in aggregation group $agg_group up: $computes_number_up_before_logic"
+  log "0" "Expected number of hypervisors in aggregation group $agg_group up: $computes_number_up_after_logic"
+  log "0" "Hypervisors in aggregation group $agg_group up: $(echo $computes_hostnames_up)"
+  log "0" "Hypervisors in aggregation group $agg_group down: $(echo $computes_hostnames_down_with_no_vm)"
+}
+
+computes_shutdown() {
```

Comment:
computes_shutdown and computes_startup functions are more or less the same code copy pasted, move this into a single function to poweron/poweoff as I said earlier. You only have to send to this function the list of computes and if you want to poweoff/poweon those. Also format the log strings from this function accordingly.